### PR TITLE
WIP: SELinux checking overhaul and logind addition

### PR DESCRIPTION
**Pull request by cgzones**
CI test

### CODE REVIEW

File: src/shared/unit-file.c [hunk 1]
```diff
@@ -385,10 +394,27 @@ int unit_file_build_name_map(
...
+                                        set_remove(masked_ids, de->d_name);
+                                }
+                        } else {
+                                r = hashmap_put_strdup(&ids, de->d_name, dst);
+                                if (r < 0)
+                                        return log_warning_errno(r, "Failed to add entry to hashmap (%s→%s): %m",
+                                                                de->d_name, dst);
+
+                                if (null_or_empty_path(dst)) {
+                                        r = set_put_strdup(&masked_ids, de->d_name);
```
Comment (mrc0mmand) [inline src/shared/unit-file.c]:
```
==1==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 141 byte(s) in 5 object(s) allocated from:
    #0 0x7f47220bd8e6 in __interceptor_strdup /build/gcc/src/gcc/libsanitizer/asan/asan_interceptors.cc:445
    #1 0x7f4721550318 in set_put_strdup ../build/src/basic/hashmap.c:1818
    #2 0x7f4721490b29 in unit_file_build_name_map ../build/src/shared/unit-file.c:413
    #3 0x56138a1c16bf in unit_load_fragment ../build/src/core/load-fragment.c:4765
    #4 0x56138a2c2a1f in unit_load_fragment_and_dropin ../build/src/core/unit.c:1461
    #5 0x56138a3af346 in mount_load ../build/src/core/mount.c:649
    #6 0x56138a2daa5c in unit_load ../build/src/core/unit.c:1636
    #7 0x56138a1dd041 in manager_dispatch_load_queue ../build/src/core/manager.c:1953
    #8 0x56138a1dd571 in manager_enumerate ../build/src/core/manager.c:1467
    #9 0x56138a1ee954 in manager_startup ../build/src/core/manager.c:1679
    #10 0x56138a0fe804 in main ../build/src/core/main.c:2713
    #11 0x7f4720062022 in __libc_start_main (/usr/lib/libc.so.6+0x27022)

SUMMARY: AddressSanitizer: 141 byte(s) leaked in 5 allocation(s).
```

**Comment by mrc0mmand** [comment]
Also this from `TEST-04-JOURNAL`:

```
[  280.765689] testsuite-04.sh[353]: + touch /testok
[  OK  ] Stopped systemd-update-utmp.service.
[  OK  ] Stopped systemd-tmpfiles-setup.service.
[  OK  ] Stopped target local-fs.target.
[  OK  ] Stopped target local-fs-pre.target.
         Unmounting tmp.mount...
[  OK  ] Stopped systemd-tmpfiles-setup-dev.service.
[  OK  ] Stopped systemd-sysusers.service.
[  OK  ] Stopped systemd-remount-fs.service.
         Starting systemd-journal-flush.service...
[  OK  ] Unmounted tmp.mount.
[  OK  ] Stopped target swap.target.
[  OK  ] Reached target shutdown.target.
[  OK  ] Reached target umount.target.
[  OK  ] Reached target final.target.
[  OK  ] Finished systemd-poweroff.service.
[  OK  ] Reached target poweroff.target.
[  286.755493] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000100
[  286.757172] CPU: 1 PID: 1 Comm: systemd Not tainted 5.6.11-arch1-1 #1
[  286.757515] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS ?-20191223_100556-anatol 04/01/2014
[  286.757916] Call Trace:
[  286.757916]  dump_stack+0x66/0x90
[  286.757916]  panic+0x112/0x2e8
[  286.757916]  do_exit.cold+0xa7/0xc1
[  286.757916]  ? handle_mm_fault+0xce/0x200
[  286.757916]  do_group_exit+0x3a/0xa0
[  286.757916]  __x64_sys_exit_group+0x14/0x20
[  286.757916]  do_syscall_64+0x4e/0x150
[  286.757916]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[  286.757916] RIP: 0033:0x7f4ec21be002
[  286.757916] Code: 00 0f 05 c3 0f 1f 40 00 f3 0f 1e fa b8 18 00 00 00 0f 05 c3 0f 1f 40 00 f3 0f 1e fa 48 83 ec 08 48 63 ff b8 e7 00 00 00 0f 05 <67> e8 c8 ee 00 00 0f 1f 84 00 00 00 00 00 f3 0f 1e fa 48 83 ec 28
[  286.757916] RSP: 002b:00007ffc88275c60 EFLAGS: 00000206 ORIG_RAX: 00000000000000e7
[  286.757916] RAX: ffffffffffffffda RBX: 00007f4ec2ac7558 RCX: 00007f4ec21be002
[  286.757916] RDX: 0000000000000001 RSI: 0000000000000000 RDI: 0000000000000001
[  286.757916] RBP: 00007f4ec2ac7558 R08: 00007f4eb539345b R09: 0000000000000000
[  286.757916] R10: 0000000000000040 R11: 0000000000000206 R12: 00000fe9d778dc00
[  286.757916] R13: 00007f4ebbc6e000 R14: 0000000000000000 R15: 000000000141364e
[  286.757916] Kernel Offset: 0x8000000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
[  286.757916] ---[ end Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000100 ]---
qemu-system-x86_64: terminating on signal 15 from pid 107270 (timeout)
```

File: src/shared/unit-file.c (hunk 1 above)
Comment (mrc0mmand) [reply src/shared/unit-file.c]: Another one popped up:

```
==1==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 435 byte(s) in 17 object(s) allocated from:
    #0 0x7f67fcf8f8e6 in __interceptor_strdup /build/gcc/src/gcc/libsanitizer/asan/asan_interceptors.cc:445
    #1 0x7f67fc41b316 in set_put_strdup ../build/src/basic/hashmap.c:1818
    #2 0x7f67fc35bb27 in unit_file_build_name_map ../build/src/shared/unit-file.c:414
    #3 0x562b7b89b6bf in unit_load_fragment ../build/src/core/load-fragment.c:4765
    #4 0x562b7b99ca1f in unit_load_fragment_and_dropin ../build/src/core/unit.c:1461
    #5 0x562b7ba89346 in mount_load ../build/src/core/mount.c:649
    #6 0x562b7b9b4a5c in unit_load ../build/src/core/unit.c:1636
    #7 0x562b7b8b7041 in manager_dispatch_load_queue ../build/src/core/manager.c:1953
    #8 0x562b7b8b7571 in manager_enumerate ../build/src/core/manager.c:1467
    #9 0x562b7b8c91d6 in manager_reload ../build/src/core/manager.c:3596
    #10 0x562b7b7d52f1 in invoke_main_loop ../build/src/core/main.c:1796
    #11 0x562b7b7d8958 in main ../build/src/core/main.c:2741
    #12 0x7f67faf07022 in __libc_start_main (/usr/lib/libc.so.6+0x27022)

Direct leak of 435 byte(s) in 17 object(s) allocated from:
    #0 0x7f67fcf8f8e6 in __interceptor_strdup /build/gcc/src/gcc/libsanitizer/asan/asan_interceptors.cc:445
    #1 0x7f67fc41b316 in set_put_strdup ../build/src/basic/hashmap.c:1818
    #2 0x7f67fc35bb27 in unit_file_build_name_map ../build/src/shared/unit-file.c:414
    #3 0x562b7b89b6bf in unit_load_fragment ../build/src/core/load-fragment.c:4765
    #4 0x562b7b99ca1f in unit_load_fragment_and_dropin ../build/src/core/unit.c:1461
    #5 0x562b7ba89346 in mount_load ../build/src/core/mount.c:649
    #6 0x562b7b9b4a5c in unit_load ../build/src/core/unit.c:1636
    #7 0x562b7b8b7041 in manager_dispatch_load_queue ../build/src/core/manager.c:1953
    #8 0x562b7b8b7571 in manager_enumerate ../build/src/core/manager.c:1467
    #9 0x562b7b8c8954 in manager_startup ../build/src/core/manager.c:1679
    #10 0x562b7b7d8804 in main ../build/src/core/main.c:2713
    #11 0x7f67faf07022 in __libc_start_main (/usr/lib/libc.so.6+0x27022)
```

**Comment by poettering** [comment]
How does this relate to #10023? It's the same stuff? Can one be closed?

**Comment by cgzones** [comment]
Yes, this was a WIP version of 10023.
I opened a new pull request for a CI run and to not pollute the original pr.
I am going to refresh the original pull request shortly but I'd like to test things here first.

**Comment by evverx** [comment]
> I opened a new pull request for a CI run and to not pollute the original pr.

The "selinux" part has no tests whatsover (apart from `test-selinux` (which doesn't really test anything)) so the CI can only make sure that systemd compiles with libselinux enabled/disabled (which is kind of helpful but far from ideal). I wonder if before a major overhaul like this it would be better to get some tests working just to make sure the part nobody seems to have complained about isn't going to be broken.

**Comment by evverx** [comment]
Interestingly, it seems `libselinux` itself has no tests whatsoever and to judge from https://github.com/SELinuxProject/selinux/issues/249 crashes as soon as anyone tries to run it with ASan. I've just tried to build SELinuxProject/selinux on Fedora and it (unsurprisingly) failed to compile with
```
direct_api.c: In function ‘semanage_direct_commit’:
direct_api.c:1466:13: error: implicit declaration of function ‘sepol_policydb_optimize’; did you mean ‘sepol_policydb_to_image’? [-Werror=implicit-function-declaration]
 1466 |    retval = sepol_policydb_optimize(out);
      |             ^~~~~~~~~~~~~~~~~~~~~~~
      |             sepol_policydb_to_image
cc1: all warnings being treated as errors
make[2]: *** [Makefile:110: direct_api.o] Error 1
```
Given that it seems even the basics aren't covered there it looks like it's going to be hard to somehow bring it to our CI.

**Comment by cgzones** [comment]
The ASAN crash seems to be fixed in 3.1-rc2.
The compile error seems to result from not compiling libsemanage against an adequate libsepol version. Building all the sub-projects of SELinuxProject/selinux requires building against the in-repository dependencies, either via overriding include and lib directories or building and installing each single project.

The SELinux part of systemd must indeed be tested by-hand (by setting up an own vm and play around with the access permission required actions).

**Comment by evverx** [comment]
> The ASAN crash seems to be fixed in 3.1-rc2.

It was fixed. Though I can't seem to find a place where `libselinux` would be tested regularly with ASan to make sure that issues like that wouldn't pop up again a few versions later.

> The compile error seems to result from not compiling libsemanage against an adequate libsepol version. Building all the sub-projects of SELinuxProject/selinux requires building against the in-repository dependencies, either via overriding include and lib directories or building and installing each single project.

That's usually what build systems are for. Manually overriding the defaults or checking the build dependencies isn't what I would expect from a CI-friendly library.

> The SELinux part of systemd must indeed be tested by-hand (by setting up an own vm and play around with the access permission required actions).

Another option would be to resurrect https://github.com/systemd/systemd/tree/master/test/TEST-06-SELINUX where VM is set up automatically)

**Comment by evverx** [comment]
@cgzones thank for the pointers. I managed to build the project with ASan+UBsan and ran the unit tests: https://github.com/SELinuxProject/selinux/issues/253.

**Pull request closed without merging**